### ISSUE TITLE
My trigger schedule doesn't seem to be honored (v0.10.0)

### ISSUE THREAD

**Issue by samip5** [backube/volsync#1415]
**Describe the bug**
I have specified a trigger schedule to backup at 2AM, but it seems it's not honored?

**Steps to reproduce**
Set a replicationsource with trigger schedule of `0 2 * * *`

**Expected behavior**
I expected it to honor my trigger.

**Actual results**
It doesn't seem to honor it.

**Additional context**

I'm using Volume Populator, and my replication source is here:
```
apiVersion: volsync.backube/v1alpha1
kind: ReplicationSource
metadata:
  creationTimestamp: "2024-07-28T14:35:32Z"
  generation: 2
  labels:
    kustomize.toolkit.fluxcd.io/name: dev-forego
    kustomize.toolkit.fluxcd.io/namespace: flux-system
  name: forego-r2
  namespace: dev
  resourceVersion: "221092293"
  uid: 6107f4d4-56a4-499a-aaea-f66511200091
spec:
  restic:
    cacheCapacity: 15Gi
    cacheStorageClassName: zfs
    copyMethod: Snapshot
    moverSecurityContext:
      fsGroup: 1000
      runAsGroup: 1000
      runAsUser: 1000
    pruneIntervalDays: 7
    repository: forego-volsync-r2
    retain:
      daily: 7
      weekly: 1
    storageClassName: zfs
    volumeSnapshotClassName: zfs
  sourcePVC: forego-data
  trigger:
    schedule: 0 2 * * *
```

Logs that show amount of snapshots: https://gist.github.com/samip5/b7a9bdf8345a5957cb659718812ffd6c

It seems that for whatever reason, it's not following the trigger, like you can clearly see from:
```
3840aa4c  2024-10-10 00:03:46  volsync                 /data  500.240 MiB
1fa0a425  2024-10-10 00:10:02  volsync                 /data  500.240 MiB
629bf12a  2024-10-10 00:17:47  volsync                 /data  500.240 MiB
c5de5a39  2024-10-10 00:28:07  volsync                 /data  500.240 MiB
```

So why would that be?

**Comment by tesshuflower** [comment]
@samip5 Is the replication completing successfully?  I can't tell from the container log.  Do you have the full replicationsource (including the status) from your system?  It should update the `.status.nextSyncTime` which should tell you when it thinks the next sync should be.

**Comment by samip5** [comment]
It would seem to have been the case which is somewhat an frequently occurring theme, that it gets to a state where it's locked by an older backup job (which is nowhere to be found again) and requires manual intervention to resolve.

**Comment by samip5** [comment]
> > It would seem to have been the case which is somewhat an frequently occurring theme, that it gets to a state where it's locked by an older backup job (which is nowhere to be found again) and requires manual intervention to resolve.
> 
> I'm not sure I understand this part - is the repo getting locked after a failed job? Is the mover pod getting killed while backing up (from OOM or something else?). If the backup doesn't complete due to an error, volsync will keep retrying, which potentially is why you see it keep attempting.

I really wish I knew how it happens (but it does require me to run unlock), as the pod is nowhere to be found, even if it failed, so I cannot figure it out.. It however results in many pods when it fails and volsync keeps on trying and trying, but fails due to the lock, and before we know it there's 5+ pods trying to do it, and as the backup itself is not instant, it takes more than 3 minutes to do... The plot is thickened.

**Comment by tesshuflower** [comment]
VolSync's mover uses a job, and should only be running 1 pod at a time. You're mentioning unlock, do you have logs from any of the failed pods indicating the lock is an issue?

If you need to run an unlock, there is an unlock option now for a restic job you could look at - see the `unlock` backup option here: https://github.com/backube/volsync/issues/1421

**Comment by onedr0p** [comment]
> VolSync's mover uses a job, and should only be running 1 pod at a time.

I think what @samip5 is saying is that when the job fails, a new pod will take it's place, fail and stay in an errored state, another new one will take it's place, fail stay in an errored state, etc.., etc.. New pods will be created until the end of time until the restic repo is in a healthy state without any locks. Once there is a successful backup all the errored pods will be deleted as the job is cleaned up by volsync.

This also makes the unlock problem a bit tedious to deal with. IIRC you need to pass the `--remove-all` flag to the restic unlock command to clean up all the locks on the repo since all previous tries are saved in the restic repo and locked. From what I can tell is the current way the `unlock` works with volsync is that is only unlocks the latest attempt but there could still be locks set on previous tries within the restic repo. It looks like that was previously experimented with as I see in [this](https://github.com/backube/volsync/blob/6115061aa40faa4e53d85e01bf54260585737f9e/mover-restic/entry.sh#L169) comment.

**Comment by onedr0p** [comment]
Unrelated but I find it pretty rad that [rustic](https://github.com/rustic-rs/rustic) is lock free.

- https://rustic.cli.rs/docs/comparison-restic.html
- https://github.com/restic/restic/issues/1141#issuecomment-1112621264

**Comment by samip5** [comment]
Yes, that's what I was trying to say but onedr0p did a better job explaining it.

**Comment by onedr0p** [comment]
@tesshuflower I wonder if we could make use of the [pod failure policy](https://kubernetes.io/docs/concepts/workloads/controllers/job/#pod-failure-policy) feature and [restics exit codes](https://restic.readthedocs.io/en/latest/075_scripting.html#exit-codes) for a locked repo?

e.g. set this on the job...

```yaml
podFailurePolicy:
  rules:
    - action: FailJob
      onExitCodes:
        containerName: restic
        operator: In
        values: [11]
```

**Comment by tesshuflower** [comment]
I think we never passed in the restic `--remove-all` flag as it seemed dangerous - this would also remove locks that could be set by someone else who's accessing the restic repo at the same time.

@samip5 I'd suggest to debug you could run with a manual schedule and look at the pod logs to see if there is any more detail. A restic lock doesn't normally get left behind if restic exits normally, so sounds like a crash of some sort or possibly the pod itself getting killed for OOM or some other reason.  Another possibility I suppose is disk space issues on the repo or the cache.

To debug in more detail, you could try adding the following annotation to the ReplicationSource:

`volsync.backube/enable-debug-mover` (with any value, even "").

When the next synchronization runs (you can run a manual trigger to run it immediately), the pod will now sit in "debug" mode (it will be idle and not run any backup), and then you can exec into it.  The pod log has instructions but will have copied the mover script to /tmp where you can run it.  You could then manually run the script which will invoke the backup and see if you get any more details.  You can even modify the script in /tmp before running it and add `-vvv` to the restic backup command.

**Comment by samip5** [comment]
Hit this again, and yes it's the lock that's the problem in 99.9% of my cases at least. I don't think the debug mover would show anything more about it.

I don't really care why it gets left behind, I only care that it does which prevents further backups.
